mmseg/models/decode_heads/hrnetconstrast_head.py

```diff
@@ -0,0 +1,174 @@
+            nn.Dropout2d(drop_prob),
+            nn.Conv2d(in_channels, out_channels, kernel_size=1),
+        )
+
+    def forward(self, x):
+        return self.seg(x)
+
+
+@MODELS.register_module()
+class HRNetContrastHead(BaseDecodeHead):
+    """The segmentation head used by contrast learning.
+
+    Args:
+        drop_p (float):
+            The probability of dropout in segment head.
+        proj_n (int):
+            Each pixel will be projected into a vector with length of proj_n.
+        proj_mode (str):
+            The mode for project head ,'linear' or 'convmlp'.
+    """
+
+    def __init__(self, drop_p=0.1, proj_n=256, proj_mode='convmlp', **kwargs):
+        super().__init__(**kwargs)
+        if proj_n <= 0:
+            raise KeyError('proj_n must >0')
+        if drop_p < 0 or drop_p > 1 or not isinstance(drop_p, float):
+            raise KeyError('drop_p must be a float >=0')
+        self.proj_n = proj_n
+
+        self.seghead = SegmentationHead(
+            in_channels=self.in_channels,
+            out_channels=self.num_classes,
+            drop_prob=drop_p)
+        self.projhead = ProjectionHead(
+            in_channels=self.in_channels, out_channels=proj_n, proj=proj_mode)
+
+    def forward(self, inputs):
+        inputs = self._transform_inputs(inputs)
+        output = {}
+        output['seg'] = self.seghead(inputs)
+
+        output['proj'] = self.projhead(inputs)
+
+        return output
+
+    def loss_by_feat(self, seg_logits: dict,
+                     batch_data_samples: SampleList) -> dict:
+        """Compute segmentation loss.
+
+        Args:
+            seg_logits (dict): The output from decode head forward function.
+            batch_data_samples (List[:obj:`SegDataSample`]): The seg
+                data samples. It usually includes information such
+                as `metainfo` and `gt_sem_seg`.
+
+        Returns:
+            dict[str, Tensor]: a dictionary of loss components
+        """
+
+        seg_label = self._stack_batch_gt(batch_data_samples)
+        loss = dict()
+
+        if self.sampler is not None:
+            seg_weight = self.sampler.sample(seg_logits['seg'], seg_label)
+        else:
+            seg_weight = None
+        seg_label = seg_label.squeeze(1)
+
+        if not isinstance(self.loss_decode, nn.ModuleList):
+            losses_decode = [self.loss_decode]
+        else:
+            losses_decode = self.loss_decode
+        for loss_decode in losses_decode:
+            if loss_decode.loss_name not in loss:
+                loss[loss_decode.loss_name] = loss_decode(
+                    seg_logits,
+                    seg_label,
+                    weight=seg_weight,
+                    ignore_index=self.ignore_index)
+            else:
+                loss[loss_decode.loss_name] += loss_decode(
+                    seg_logits,
+                    seg_label,
+                    weight=seg_weight,
+                    ignore_index=self.ignore_index)
```

Comment:
Following the implementation of PaddleSeg, the input of different loss methods should be different.

https://github.com/PaddlePaddle/PaddleSeg/blob/2c8c35a8949fef74599f5ec557d340a14415f20d/paddleseg/core/train.py#L39